Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java

```diff
@@ -28,44 +28,224 @@
+     * determined by locationList[i] or (for directLocationMapping) the point at
+     * store[i * dimensions]
+     */
+    protected short[] refCount;
+    /**
+     * first location where new data can be safely copied;
+     */
+    int startOfFreeSegment;
+    /**
+     * overall dimension of the point (after shingling)
+     */
+    int dimensions;
+    /**
+     * shingle size, if known. Setting shingle size = 1 rules out overlapping
+     */
+    int shingleSize;
+    /**
+     * number of original dimensions which are shingled to produce and overall point
+     * dimensions = shingleSize * baseDimensions. However there is a possibility
+     * that even though the data is shingled, we may not choose to use the
+     * overlapping (say for out of order updates).
+     */
+    int baseDimension;
+    /**
+     * even for shingleSize>1 we may seek to disable certain behavior, for example
+     * if we know that the set of points in the store already have low overlap or
+     * the order of updates is not in a sequential increasing order
+     */
+    boolean shingleAwareOverlapping;
+    /**
+     * are the addresses mapped directly (saves space and runtime for shingleSize =
+     * 1)
+     */
+    boolean directLocationMap;
 
     /**
      * Create a new PointStore with the given dimensions and capacity.
      *
      * @param dimensions The number of dimensions in stored points.
      * @param capacity   The maximum number of points that can be stored.
      */
+    public PointStore(int dimensions, int shingleSize, int capacity, boolean shingleAwareOverlapping,
+            boolean directLocationMap) {
+        super(capacity);
+        checkArgument(dimensions > 0, "dimensions must be greater than 0");
+        checkArgument(shingleSize == 1 || dimensions % shingleSize == 0, "incorrect use");
+        checkArgument(!directLocationMap || !shingleAwareOverlapping,
+                " cannot have overlapped shingles and direct map simultaneously");
+        this.shingleSize = shingleSize;
+        this.dimensions = dimensions;
+        refCount = new short[capacity];
+        startOfFreeSegment = 0;
+        this.directLocationMap = directLocationMap;
+        this.shingleAwareOverlapping = shingleAwareOverlapping;
+        if (!directLocationMap) {
+            if (shingleAwareOverlapping && shingleSize > 1) {
+                // even if shingle size is 1
+                baseDimension = dimensions / shingleSize;
+            } else {
+                baseDimension = dimensions;
+            }
+            locationList = new int[capacity];
+            if (!shingleAwareOverlapping) { // initialize a 1-1 map
+                for (int i = 0; i < capacity; i++) {
+                    locationList[i] = i * dimensions;
+                }
+            }
+        } else {
+            baseDimension = dimensions;
+        }
+    }
+
     public PointStore(int dimensions, int capacity) {
```

Comment:
Could you re-use the other constructor for this. E.g., is this the same as
```
public PointStore(int dimensions, int capacity) {
  this(dimensions, 1, capacity, false, true);
}
```

Reply:
done